Improve user symbol resolution

lenticularis39 opened the pull request:
<!--
Please provide a description of your change below this comment.

Then please complete the checklist.
-->
This PR contains three related improvements of usym resolution (edit: added purpose of the changes):
- Caching BCC symcache per process if ASLR is enabled (edit: optional after changes, by default disabled)
  - Enables faster symbol resolution and symcache preloading (see below)
- Resolving symbols directly from ELF (only if ASLR is disabled)
  - Enables resolving symbols that are not dynamically loaded even after process exit
- Preloading BCC symcache for running instances of program
  - Enables resolving symbols after program exists even with ASLR enabled, given that the process exists at probe attachment time.

The latter two provide a partial fix of #2284. 

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

viktormalik commented:
CI failed, can you debug it?

lenticularis39 commented:
I looked into that, one issue is codegen tests, because there is now one more attribute of the usym event (that one is easily fixable), other issue is ustack runtime tests failing with segfault (that one will be a little bit more difficult, since it doesn't happen on my system).

Also there is another problem with the PR, that is that the necessary BCC fix was reverted in iovisor/bcc#4270 (it turned that `/proc/.../exe` is not a simple symlink, but it also reflects mount namespaces, meaning the patch broke symbol resultion for processes running in a different namespace). That has to be resolved in order for the last of this PR to work.

viktormalik reviewed (commented):
Overall, these are very nice changes!

What I'm missing here, though, are better explanations of *why* the given approaches are necessary - this should appear in the PR description, in commit messages, and in code comments, too. Perhaps even a single-place summary of all the considered scenarios and corresponding solutions could be useful.

lenticularis39 commented on src/bpftrace.cpp:
```diff
@@ -2129,23 +2157,40 @@ bool BPFtrace::is_aslr_enabled(int pid)
+      if (!pid_exe.empty() && !show_offset && !show_module)
+      {
+        // try to resolve symbol directly from binary
+        std::map<uint64_t, std::string> &symbol_table =
+            symbol_table_cache_.find(pid_exe) != symbol_table_cache_.end()
+                ? symbol_table_cache_[pid_exe]
+                : (symbol_table_cache_[pid_exe] = get_symbol_table_for_elf(
+                       pid_exe));
```
The issue with using insert is that under eager evaluation this always calls `get_symbol_table_for_elf`, making it less efficient.

viktormalik replied:
Ah, right. This looks ugly, but seems like we don't have much of a better option.
Perhaps this could be an alternative, but I'm not strongly against leaving the original one:
```suggestion
        if (symbol_table_cache_.find(pid_exe) == symbol_table_cache_.end())
            symbol_table_cache_[pid_exe] = get_symbol_table_for_elf(pid_exe)
        auto &symbol_table = symbol_table_cache_[pid_exe];
```

lenticularis39 commented:
Rebased and changed caching by PID, which is required for per-process symbol cache preloading, to optional (see the last commit for details).

viktormalik commented on tests/runtime/builtin:
```diff
@@ -108,6 +108,13 @@ EXPECT SUCCESS .*
+#PROG uprobe:./testprogs/uprobe_symres_exited_process:test { print(func); exit(); }
+#EXPECT test
+#BEFORE ./testprogs/uprobe_symres_exited_process
+#TIMEOUT 5
```
For other tests, we use
```
REQUIRES bash -c "exit 1"
```
to explicitly disable the test.

lenticularis39 replied:
Ah, nice, I was looking for a way to disable a test globally and I didn't find anything.

viktormalik reviewed (approved):
Looks good but it's quite a complex change so I'd like to have another review before merging.

ajor commented on man/adoc/bpftrace.adoc:
```diff
@@ -1761,8 +1765,9 @@ The pointer type is left unchanged.
-Equal to <<functions_ksym>> but resolves user space symbols
+Equal to <<functions_ksym>> but resolves user space symbols.
 
+If ASLR is enabled, it works only when the process is running at either the time of the symbol resolution or the time of the probe attachment. The latter requires `BPFTRACE_CACHE_USER_SYMBOLS` to be set to `PER_PID`, and might not work with older versions of BCC. A similar limitation also applies to dynamically loaded symbols.
```
```suggestion
If ASLR is enabled, user space symbolication only works when the process is running at either the time of the symbol resolution or the time of the probe attachment. The latter requires `BPFTRACE_CACHE_USER_SYMBOLS` to be set to `PER_PID`, and might not work with older versions of BCC. A similar limitation also applies to dynamically loaded symbols.
```
Nit: Reworded to sound better to me

ajor commented on src/bpftrace.cpp:
```diff
@@ -2196,6 +2197,11 @@ std::string BPFtrace::resolve_usym(uintptr_t addr,
+    else
+    {
+      // no user symbol caching, create new bcc cache
+      psyms = bcc_symcache_new(pid, &symopts);
```
The old code had a matching `bcc_free_symcache` when caching was disabled - I assume we'll leak here without it

ajor commented on src/main.cpp:
```diff
@@ -305,21 +305,44 @@ static std::optional<struct timespec> get_boottime()
+  {
+    std::string s(env_p);
+    // Note: options 0 and 1 are for compatibility with older versions of
+    // BPFtrace
```
```suggestion
    // bpftrace
```
The old name really refuses to go away! 😂

ajor commented on src/main.cpp:
```diff
@@ -305,22 +305,44 @@ static std::optional<struct timespec> get_boottime()
+    }
+  }
 
   if (!std::getenv("BPFTRACE_CACHE_USER_SYMBOLS"))
```
Can this just be an "else" or am I missing something?

ajor commented on man/adoc/bpftrace.adoc:
```diff
@@ -210,12 +210,16 @@ Increasing the value will consume more memory, increase startup times and can in
 === BPFTRACE_CACHE_USER_SYMBOLS
 
-Default: 0 if ASLR is enabled on system and `-c` option is not given; otherwise 1
+Default: PER_PROGRAM if ASLR disabled, PER_PID if ASLR enabled and -p option used; NONE otherwise.
```
This doesn't match the implemented behaviour, although the way it's described here sounds good to me

lenticularis39 replied:
I don't see any difference between the behavior described here and the behavior implemented in `src/main.cpp`. Anyway, I'll change both to reflect the suggestion to default to PER_PID.

ajor commented:
I love this PR, thanks so much for doing it - it solves so many problems I've had with user symbols over the years!

I'd suggest that we make "PER_PID" the default caching behaviour when ASLR is enabled, and leave it up to users to disable the cache if they feel it's using too much memory.
1. Having ASLR enabled is not too uncommon, and often users will only be tracing a single process. I'd think the large amount of forking described in #816 is a less common scenario
2. It's simpler for users to understand disabling a cache than it is to work out which type of caching to enable

ajor commented on src/bpftrace.cpp:
```diff
@@ -2099,23 +2134,44 @@ bool BPFtrace::is_aslr_enabled(int pid)
+      auto probe_full = resolve_probe(probe_id);
+      size_t start = probe_full.find(':') + 1;
+      size_t end = probe_full.find(':', start);
+      pid_exe = probe_full.substr(start, end - start);
```
Is the caching just expected to fail when using wildcards in probes, e.g. "uprobe:/bin/gc*:*"

A comment explaining the expected behaviour would be useful for future developers

lenticularis39 replied:
Yes, since the program is not expanded, we don't know which executable the perf event came from if a wildcard is used. I'll add a comment explaining this.

ajor commented on src/ast/passes/resource_analyser.cpp:
```diff
@@ -222,6 +236,11 @@ void ResourceAnalyser::prepare_mapped_printf_ids()
 
+bool ResourceAnalyser::uses_usym_table(const std::string &fun)
+{
+  return fun == "usym" || fun == "func" || fun == "ustack";
```
ustack is listed here, but it doesn't use the usym table, right?

lenticularis39 replied:
`ustack` uses `resolve_usym` to resolve addresses in the stack to symbols, as shown in your comment above, or did you mean something else?

ajor commented on src/required_resources.h:
```diff
@@ -116,6 +116,9 @@ class RequiredResources
   std::vector<Probe> watchpoint_probes;
 
+  // List of probes using userspace symbol resolution
+  std::unordered_set<std::string> probes_using_usym;
```
```suggestion
  std::unordered_set<std::reference_wrapper<Probe>> probes_using_usym;
```
Storing a reference will be a bit faster and use less memory than basing this set on strings. Theoretically we could also get probe name collisions, which references would prevent.

lenticularis39 commented on src/bpftrace.cpp:
```diff
@@ -303,6 +315,29 @@ int BPFtrace::add_probe(ast::Probe &p)
+        // this allows symbol resolution for processes that are running at probe
+        // attach time, but not at symbol resolution time, even with ASLR
+        // enabled, since BCC symcache records the offsets
+        for (int pid : get_pids_for_program(attach_point->target))
```
It shouldn't be too much, e.g. `ls -l /proc/*/exe` takes 20ms on a system with around 350 processes, and the iteration only happens once per probe (which can be further reduced to just once with a better implementation).

lenticularis39 commented:
Changed default setting of `BPFTRACE_CACHE_USER_SYMBOLS` to `PER_PROGRAM` without ASLR and `PER_PID` per ASLR, updated documentation, resolved comments, rebased to master and updated changelog with the new changes.

ajor commented:
My comments about "ustack" come from it not passing a "probe_id" value back from the kernel like "usym" now does. The code is currently passing in "-1" as a probe_id for ustacks, which means they don't get all the features of usym (post-termination symbol resolution).

On another note: I'm wondering if we need to do some probe expansion when using usym now, i.e. similar to `probe_->need_expansion`.

I assume that would solve the wildcard issue I commented on before, and I'm also worried that we can potentially resolve symbols incorrectly against the wrong binary as it stands.

e.g. say we have a probe:
```probe:myprog1:funcA,uprobe:myprog2:funcB { @sym = usym) }```

If the process terminates before we resolve the symbol, I believe the current code will always pick "myprog1" as the pid_exe. If it was actually an address from myprog2, then we could end up with completely the wrong symbol. (not tested, just assuming)

Expansion would only really be needed when there are multiple different exes sharing a single probe. Multiple functions from the same exe sharing a probe would not need this. Multiple-function-single-exe is going to be a much more common case, and over-expanding probes would have some significant overheads when tracing a large number of functions, so is something we should avoid.

I realise that adding in this logic would not necessarily be trivial, and the rest of your changes in this PR look great, so if you want we could merge this with the post-process-termination logic temporarily disabled and follow up with a separate PR. Up to you.

ajor commented:
Or maybe just explicitly disabling the post-process-termination logic for probes specifying multiple target paths (with either wildcards or commas) would be good enough as a permanent solution. That'd leave it working fast in the common case (probes with a single target path), and ensure we still get correct results, albeit slower, in the less common case.

brendangregg commented:
I just heard about https://github.com/libbpf/blazesym which Meta is working on. At LSFMMBPF there was talk about whether it made sense for bcc and bpftrace to use blazesym (talk as in "does someone want to explore doing this?").

lenticularis39 commented:
I disabled ELF symbol resulution for probes with multiple attach points as suggested, since we don't know which binary is the correct one. `kstack` now passed `probe_id` to `resolve_usym`; I also updated the ELF symbol resolution logic to work with addresses with non-zero offsets, and removed the code for legacy BCC in `get_symbol_table_for_elf`, since we now ship BCC as a submodule.

(The failing tests are just a missing cast, I'll fix that soon.)

viktormalik commented:
> I just heard about https://github.com/libbpf/blazesym which Meta is working on. At LSFMMBPF there was talk about whether it made sense for bcc and bpftrace to use blazesym (talk as in "does someone want to explore doing this?").

@brendangregg we were discussing this at one of the bpftrace office hours meetings. We're slowly trying to get rid of dependency on bcc and symbolization is one of those parts where we depend on it a lot. I'd keep that discussion outside of this PR, though.

ajor commented on src/bpftrace.h:
```diff
@@ -223,7 +236,11 @@ class BPFtrace
+  // note: exe_sym_ is used when layout is same for all instances of program
   std::map<std::string, std::pair<int, void *>> exe_sym_; // exe -> (pid, cache)
+  std::map<int, void *> pid_sym_;                         // pid -> cache
+  std::map<std::string, std::map<uint64_t, elf_symbol, std::greater<uint64_t>>>
```
I think uintptr_t would be a better type for storing symbols. It's what we use everywhere else at least

lenticularis39 replied:
Makes sense, on 32-bit platforms it will save some memory.

ajor commented on tests/runtime/builtin:
```diff
@@ -108,6 +108,15 @@ EXPECT SUCCESS .*
 AFTER ./testprogs/uprobe_negative_retval
 TIMEOUT 5
 
+# Disabled, since BCC code it depends on is prone to race condition
+NAME func_uprobe_exited_process
+ENV BPFTRACE_CACHE_USER_SYMBOLS=PER_PID
+PROG uprobe:./testprogs/uprobe_symres_exited_process:test { print(func); exit(); }
+EXPECT test
+BEFORE ./testprogs/uprobe_symres_exited_process
+TIMEOUT 5
+REQUIRES bash -c "exit 1"
+
```
Ok, makes sense. I think there'd be a race condition between the target process exiting and bpftrace printing, but it'd be so extremely unlikely to happen that it doesn't matter for this.

What's the race condition you mention in the comment about this test being disabled though?

lenticularis39 replied:
Just for clarification, the sequence of events that happens 99%+ of the time:
- `uprobe_symres_exited_process` starts and sleeps for 3 seconds.
- bpftrace attaches probe on `uprobe_symres_exited_process:test` and pre-loads the cache with the use of util function `get_pids_for_program` and `bcc_symcache_new`.
- the process wakes up, executes `test()` and immediately exists
- the probe fires and `bcc_symcache_resolve` is run on the stored cache (note that `symbol_table_cache_` is not used here, since `BPFTRACE_CACHE_USER_SYMBOLS` is set to `PER_PID`)

Now what happens in `bcc_symcache_resolve` is that the binary of `uprobe_symres_exited_process` is opened to look for the symbol. However, since in a general case the binary can be in a different mount namespace, BCC uses a file descriptor of `/proc/<pid>/root` that was pre-opened in `bcc_symcache_now` to access the file using `openat` (see iovisor/bcc#4319). The issue is that this approach with `openat` sometimes does not work due to current implementation issues, hence BCC checks if the process is running, and if yes, it does not use `openat`. This is prone to a race condition, since the process may exit *after* this check is done.

This race condition is much rarer than the original one from #2284 linked above (which in my case happened 100% of the time), but still happens and would cause CI randomly failing, hence the test being disabled.

ajor reviewed (commented):
> I also updated the ELF symbol resolution logic to work with addresses with non-zero offsets

Did you have an example of where this was causing symbolication problems? It'd be useful to check that in as a test-case.

I have a slight concern about the impact this pre-caching will have on startup time when tracing huge programs. I'll do a little bit of testing on this to satisfy myself.

lenticularis39 commented on src/utils.cpp:
```diff
@@ -1335,4 +1337,49 @@ std::tuple<std::string, std::string, std::string> split_addrrange_symbol_module(
                          symbol.length() - idx2 - strlen(" []")) };
 }
 
+std::map<uint64_t, elf_symbol, std::greater<uint64_t>> get_symbol_table_for_elf(
```
That is because the semantics of `std::map::lower_bound` and `std::map::upper_bound`, citing CPPreference:
- `std::map::lower_bound`: Returns an iterator pointing to the first element that is not less than (i.e. greater or equal to) key.
- `std::map::upper_bound`:  Returns an iterator pointing to the first element that is greater than key.

I want the element that is *less or equal* to the key (possible start address of the symbol), therefore I have to sort the elements of the map the other way.

lenticularis39 commented:
> > I also updated the ELF symbol resolution logic to work with addresses with non-zero offsets
> 
> Did you have an example of where this was causing symbolication problems? It'd be useful to check that in as a test-case.

The issue is in general `ustack` addresses that are not resolvable using `bcc_symcache_resolve`, but are resolvable using `symbol_table_cache_`. More specifically, it is `ustack` addresses in exited processes with ASLR disabled - I didn't write any tests for this because of the requirement of disabling ASLR (also for `usym` resolution), now when I'm thinking about it I could use `personality(ADDR_NO_RANDOMIZE)` to disable ASLR in a test program and test both.

> I have a slight concern about the impact this pre-caching will have on startup time when tracing huge programs. I'll do a little bit of testing on this to satisfy myself.

OK. The bcc symcache pre-loading is not a issue, since it only loads the symbols when `bcc_symcache_resolve` is called (since e.g. dynamically loaded modules might not be present at the time of calling `bcc_symcache_new`), what could impact the startup time is `symbol_table_cache_`, it might be useful to have an option to disable it, like we do for the symcache.

lenticularis39 commented on src/types.cpp:
```diff
@@ -119,9 +119,10 @@ bool SizedType::operator==(const SizedType &t) const
-  return type == Type::string || type == Type::usym || type == Type::inet ||
-         type == Type::buffer || type == Type::timestamp ||
-         type == Type::mac_address || type == Type::cgroup_path;
+  return type == Type::string || type == Type::usym || type == Type::ustack ||
```
Wouldn't that then also be a case for `ustack`, which is already defined as a byte array and realized with an 8-aligned struct in codegen?

ajor replied:
Good point, and I don't know...

@viktormalik Can you shed any light on this?

viktormalik replied:
So, it turns out that the `SizedType::GetAlignment` function is only used to compute alignment in tuples. I tried to change it to 8 but that causes some failures, so I'll need to investigate more.

viktormalik replied:
This is now addressed in #2625.

lenticularis39 commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -146,9 +146,43 @@ void CodegenLLVM::kstack_ustack(const std::string &ident,
     stackid = b_.CreateOr(stackid, pidhigh);
+
+    // create structure to hold both stackid and probe id
+    std::vector<llvm::Type *> elements = {
+      b_.getInt64Ty(), // stack id
+      b_.getInt64Ty(), // probe id
+    };
```
That shouldn't be too hard, I'll try to add it with improvements from the other comments, if it turns out to be too difficult to quickly implement, I'll add a comment.

lenticularis39 commented on tests/testprogs/disable_aslr.c:
```diff
@@ -0,0 +1,8 @@
+int main(int argc __attribute__((unused)), char **argv)
+{
+  personality(ADDR_NO_RANDOMIZE);
+  execv(argv[1], argv + 1);
```
I am wondering if this can be solved by an existing program, I don't know any.

viktormalik replied:
Since it's in tests, there's no need to fix this. Dismissed.

lenticularis39 commented:
Oops, I forgot to change the main offset in the test to a more generic pattern, causing the test to fail with a different compiler.

ajor reviewed (approved):
Looks great, thank you!

lenticularis39 commented:
Rebased and fixed failing test added by #2625.
